### PR TITLE
chore(enriching): check for duplicate indexes before adding a new one

### PR DESCRIPTION
**Pull request by StephenWakely**
ref #8815 

This PR checks to make sure an index hasn't already been added before adding a new one.

If it does exist, it returns a handle to the existing index.

### CODE REVIEW

**Review by spencergilbert — APPROVED** [review]
I feel like it might be good to also log an event saying "hey you've tried to index this twice"? Or is that generally not possible for a user?

File: src/enrichment_tables/file.rs [hunk 1]
```diff
@@ -211,18 +214,10 @@ impl File {
...
-    /// Creates an index with the given fields.
-    /// Uses seahash to create a hash of the data that is used as the key in a hashmap lookup to
-    /// the index of the row in the data.
-    ///
-    /// Ensure fields that are searched via a comparison are not included in the index!
-    fn index_data(
-        &self,
-        index: &[&str],
-    ) -> Result<HashMap<u64, Vec<usize>, hash_hasher::HashBuildHasher>, String> {
+    /// Order the fields in the index according to the position they are found in the header
```
Comment (pablosichert) [inline src/enrichment_tables/file.rs:217]:
```suggestion
    /// Order the fields in the index according to the position they are found in the header.
```

File: src/enrichment_tables/file.rs [hunk 2]
```diff
@@ -232,16 +227,26 @@ impl File {
...
+    }
 
+    /// Creates an index with the given fields.
+    /// Uses seahash to create a hash of the data that is used as the key in a hashmap lookup to
+    /// the index of the row in the data.
+    ///
+    /// Ensure fields that are searched via a comparison are not included in the index!
+    fn index_data(
+        &self,
+        fieldidx: &[usize],
```
Comment (pablosichert) [inline src/enrichment_tables/file.rs:240]:
```suggestion
        field_idx: &[usize],
```

File: src/enrichment_tables/file.rs (hunk 1 above)
Comment (pablosichert) [inline src/enrichment_tables/file.rs:217]:
```suggestion
    /// Order the fields in the index according to the position they are found in the header.
```

File: src/enrichment_tables/file.rs (hunk 2 above)
Comment (pablosichert) [inline src/enrichment_tables/file.rs:240]:
```suggestion
        field_idx: &[usize],
```

File: src/enrichment_tables/file.rs [hunk 3]
```diff
@@ -232,16 +227,26 @@ impl File {
...
+    ) -> Result<HashMap<u64, Vec<usize>, hash_hasher::HashBuildHasher>, String> {
         let mut index = HashMap::with_capacity_and_hasher(
             self.data.len(),
             hash_hasher::HashBuildHasher::default(),
         );
 
         for (idx, row) in self.data.iter().enumerate() {
             let mut hash = seahash::SeaHasher::default();
-            for idx in &fieldidx {
+            for idx in fieldidx {
```
Comment (pablosichert) [inline src/enrichment_tables/file.rs:249]:
```suggestion
            for idx in field_idx {
```

File: src/enrichment_tables/file.rs [hunk 4]
```diff
@@ -355,10 +361,20 @@ impl Table for File {
     }
 
     fn add_index(&mut self, fields: &[&str]) -> Result<IndexHandle, String> {
-        self.indexes.push(self.index_data(fields)?);
+        let normalized = self.normalize_index_fields(fields);
+        match self.indexes.iter().position(|index| index.0 == normalized) {
+            Some(pos) => {
+                // This index already exists
```
Comment (pablosichert) [inline src/enrichment_tables/file.rs:367]:
```suggestion
                // This index already exists.
```

**Review by pablosichert — COMMENTED** [review]
The usual nits 😄

**Review by pablosichert — APPROVED** [review]
Looks good 👍

**Review by jszwedko — APPROVED** [review]

**Comment by StephenWakely** [comment]
> I feel like it might be good to also log an event saying "hey you've tried to index this twice"? Or is that generally not possible for a user?

It's totally fine to try to add an index twice. It just means a VRL script is calling a search function with the same columns several times (but presumably with different values).

**Pull request merged**